### PR TITLE
[Cart] Change logic for maintaining cart after logging in

### PR DESCRIPTION
**Pull request by GSadee**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no?
| Deprecations?   | 
| Related tickets | 
| License         | MIT

Now, when the customer didn't have or had an empty cart and their had a cart in session, the one from session will be assigned to them.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Context/CustomerAndChannelBasedCartContext.php [hunk 1]
```diff
@@ -55,7 +55,7 @@ public function getCart(): OrderInterface
             throw new CartNotFoundException('Sylius was not able to find the cart, as there is no logged in user.');
         }
 
-        $cart = $this->orderRepository->findLatestCartByChannelAndCustomer($channel, $customer);
+        $cart = $this->orderRepository->findLatestNotEmptyCartByChannelAndCustomer($channel, $customer);
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Context/CustomerAndChannelBasedCartContext.php:58]: To be honest, I'm not 100% sure if this change will not be a little BC Break so I would like to mark this place for review 😃

File: src/Sylius/Behat/Client/ApiPlatformClient.php [hunk 2]
```diff
@@ -203,7 +203,8 @@ public function getLastResponse(): Response
         return $this->client->getResponse();
     }
 
-    private function request(RequestInterface $request): Response
+    private function
+    request(RequestInterface $request): Response
```
Comment (AdamKasp) [inline src/Sylius/Behat/Client/ApiPlatformClient.php:207]:
```suggestion
    private function request(RequestInterface $request): Response
```

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 3]
```diff
@@ -362,6 +380,10 @@ private function putProductToCart(ProductInterface $product, string $tokenValue,
 
     private function putProductVariantToCart(ProductVariantInterface $productVariant, string $tokenValue, int $quantity = 1): void
     {
+        if ($tokenValue === null) {
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:381]: cannot be true

File: src/Sylius/Bundle/CoreBundle/Context/CustomerAndChannelBasedCartContext.php (hunk 1 above)
Comment (lchrusciel) [reply src/Sylius/Bundle/CoreBundle/Context/CustomerAndChannelBasedCartContext.php:58]: I don't have a strong opinion, but we could leave the previous implementation and check object emptiness. On the other hand, we are saving object hydration, and it seems to be a cleaner solution. So 👍 from my side

**Review by lchrusciel — APPROVED** [review]

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 4]
```diff
@@ -357,11 +377,16 @@ private function putProductToCart(ProductInterface $product, string $tokenValue,
             'quantity' => $quantity,
         ]);
 
-        $this->cartsClient->executeCustomRequest($request);
+        $res = $this->cartsClient->executeCustomRequest($request);
+        $res = $res;
```
Comment (arti0090) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:381]: `
        $res = $res;
`
Is it needed? :D

Comment (arti0090) [reply src/Sylius/Behat/Context/Api/Shop/CartContext.php:381]: And is $res needed to be honest? :)

**Review by AdamKasp — APPROVED** [review]

**Comment by AdamKasp** [comment]
Thanks, Grzegorz! :tada:

**Pull request merged**